src/cpp/common/writer.h

```diff
@@ -80,6 +82,17 @@ class writer
   }
 
   void padding(offset_type size);
+
+  void add_metadata(std::unordered_map<std::string, std::string>& metadata)
+  {
+    m_metadata = std::move(metadata);
+  }
```

Comment:
You cannot move an lvalue reference, but you don't want to copy into argument either.  You should use rvalue reference and have to the calling code call add_metadata as `add_metadata(std::move(my_metadata))`
```suggestion
  void add_metadata(std::unordered_map<std::string, std::string>&& metadata)
  {
    m_metadata = std::move(metadata);
  }
```